### ISSUE TITLE
Screen sharing doesn't work in sway in Chrome

### ISSUE THREAD

**Issue by zsolt-donca** [emersion/xdg-desktop-portal-wlr#65]
Screen sharing doesn't work in any browser on my Arch Linux installation. I tried lots of browsers, wayland firefox, xwayland firefox, wayland chrome dev (the official chrome dev edition, ozone enabled with the appropriate flags), xwayland chromium (both the chrome dev and Arch Linux's chromium 86). It either just shows a black screen with my mouse cursor in it (in Firefox), or doesn't show anything at all (in chromium). Chromium has the flag _WebRTC PipeWire support_ enabled. In both cases, the mouse starts to notice and quite annoying lag (as if my screen had a 30 Hz refresh rate) until I disable the sharing.

**UPDATE:**

I just realized that I was using the default firefox which is not compatible according to the guide. Using fedora-firefox-wayland-bin works, though, though the mouse lag is still there. Thus this issue is limited to Chrome.

I have in my `~/.zshrc` the following set:
```
systemctl --user import-environment
export XDG_CURRENT_DESKTOP=sway
export XDG_SESSION_TYPE=wayland
```

I also added an `~/.zshrc.local` file:
```
XDG_CURRENT_DESKTOP=sway
XDG_SESSION_TYPE=wayland
```

And I also have an `~/.config/environment.d/sway.conf` file:
```
XDG_CURRENT_DESKTOP=sway
XDG_SESSION_TYPE=wayland
```

Thinking that my env variables are somehow still not picked up by systemd, I also tried killing all `xdg-desktop-portal`-related processes, and started everything from a terminal (where I had the variables set) with the following:
```
/usr/lib/xdg-desktop-portal -r & /usr/lib/xdg-desktop-portal-wlr
```

Nothing seems to work.

The versions of my dependencies:
```
sway 1:1.5-1
xdg-desktop-portal 1.8.0-1
xdg-desktop-portal-gtk 1.8.0-1
xdg-desktop-portal-wlr 0.1.0-5
xorg-server-xwayland 1.20.9-2
```

I am on intel GPU.

I also captured the output of `/usr/lib/xdg-desktop-portal -r & /usr/lib/xdg-desktop-portal-wlr --loglevel=DEBUG` when trying with chromium 88 with the ozone backend:

```
2020/11/08 20:05:37 [DEBUG] - dbus: connected
2020/11/08 20:05:37 [DEBUG] - wlroots: wl_display connected
2020/11/08 20:05:37 [DEBUG] - pipewire: pw_loop created
2020/11/08 20:05:37 [DEBUG] - pipewire: establishing connection to core
2020/11/08 20:05:37 [DEBUG] - wayland: registry listeners run
2020/11/08 20:05:37 [DEBUG] - wayland: xdg output listeners run
2020/11/08 20:05:51 [INFO] - dbus: create session method invoked
2020/11/08 20:05:51 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc1801729335
2020/11/08 20:05:51 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session1451295713
2020/11/08 20:05:51 [INFO] - dbus: app_id: 
2020/11/08 20:05:51 [INFO] - dbus: create session method invoked
2020/11/08 20:05:51 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc1260614061
2020/11/08 20:05:51 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session63919926
2020/11/08 20:05:51 [INFO] - dbus: app_id: 
2020/11/08 20:05:51 [INFO] - dbus: select sources method invoked
2020/11/08 20:05:51 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc530748179
2020/11/08 20:05:51 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session1451295713
2020/11/08 20:05:51 [INFO] - dbus: app_id: 
2020/11/08 20:05:51 [INFO] - dbus: option types:1
2020/11/08 20:05:51 [INFO] - dbus: option multiple: 0
2020/11/08 20:05:51 [DEBUG] - dbus: select sources: found matching session /org/freedesktop/portal/desktop/session/1_133/webrtc_session1451295713
2020/11/08 20:05:51 [INFO] - wlroots: capturable output: Sharp Corporation model: 0x1476: id: 36 name: eDP-1
2020/11/08 20:05:51 [INFO] - xdpw: screencast instance 0x558fbc979000 has 1 references
2020/11/08 20:05:51 [INFO] - xdpw: 1 active screencast instances
2020/11/08 20:05:51 [INFO] - wlroots: output: eDP-1
2020/11/08 20:05:51 [INFO] - dbus: select sources method invoked
2020/11/08 20:05:51 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc855667896
2020/11/08 20:05:51 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session63919926
2020/11/08 20:05:51 [INFO] - dbus: app_id: 
2020/11/08 20:05:51 [INFO] - dbus: option types:2
2020/11/08 20:05:51 [INFO] - dbus: option multiple: 0
2020/11/08 20:05:51 [DEBUG] - dbus: select sources: found matching session /org/freedesktop/portal/desktop/session/1_133/webrtc_session63919926
2020/11/08 20:05:51 [INFO] - wlroots: capturable output: Sharp Corporation model: 0x1476: id: 36 name: eDP-1
2020/11/08 20:05:51 [INFO] - xdpw: existing screencast instance: 36 with cursor
2020/11/08 20:05:51 [INFO] - xdpw: screencast instance 0x558fbc979000 now has 2 references
2020/11/08 20:05:51 [INFO] - wlroots: output: eDP-1
2020/11/08 20:05:51 [INFO] - dbus: start method invoked
2020/11/08 20:05:51 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc1675031948
2020/11/08 20:05:51 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session63919926
2020/11/08 20:05:51 [INFO] - dbus: app_id: 
2020/11/08 20:05:51 [INFO] - dbus: parent_window: 
2020/11/08 20:05:51 [DEBUG] - dbus: start: found matching session /org/freedesktop/portal/desktop/session/1_133/webrtc_session63919926
2020/11/08 20:05:51 [DEBUG] - pipewire: registered event 0x558fbc97a920
2020/11/08 20:05:51 [INFO] - pipewire: stream state changed to "connecting"
2020/11/08 20:05:51 [INFO] - pipewire: node id is -1
2020/11/08 20:05:51 [INFO] - dbus: start method invoked
2020/11/08 20:05:51 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc962315576
2020/11/08 20:05:51 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session1451295713
2020/11/08 20:05:51 [INFO] - dbus: app_id: 
2020/11/08 20:05:51 [INFO] - dbus: parent_window: 
2020/11/08 20:05:51 [DEBUG] - dbus: start: found matching session /org/freedesktop/portal/desktop/session/1_133/webrtc_session1451295713
2020/11/08 20:05:51 [INFO] - pipewire: stream state changed to "paused"
2020/11/08 20:05:51 [INFO] - pipewire: node id is 30
2020/11/08 20:05:52 [INFO] - dbus: session closed
2020/11/08 20:05:52 [DEBUG] - dbus: destroying session 0x558fbc979cd0
2020/11/08 20:05:52 [DEBUG] - xdpw: screencast instance 0x558fbc979000 now has 1 references
2020/11/08 20:05:52 [INFO] - dbus: session closed
2020/11/08 20:05:52 [DEBUG] - dbus: destroying session 0x558fbc978000
2020/11/08 20:05:52 [DEBUG] - xdpw: screencast instance 0x558fbc979000 now has 0 references
2020/11/08 20:05:52 [INFO] - dbus: create session method invoked
2020/11/08 20:05:52 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc1769160271
2020/11/08 20:05:52 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session1750199713
2020/11/08 20:05:52 [INFO] - dbus: app_id: 
2020/11/08 20:05:52 [INFO] - dbus: select sources method invoked
2020/11/08 20:05:52 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc1673410856
2020/11/08 20:05:52 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session1750199713
2020/11/08 20:05:52 [INFO] - dbus: app_id: 
2020/11/08 20:05:52 [INFO] - dbus: option types:1
2020/11/08 20:05:52 [INFO] - dbus: option multiple: 0
2020/11/08 20:05:52 [DEBUG] - dbus: select sources: found matching session /org/freedesktop/portal/desktop/session/1_133/webrtc_session1750199713
2020/11/08 20:05:52 [INFO] - wlroots: capturable output: Sharp Corporation model: 0x1476: id: 36 name: eDP-1
2020/11/08 20:05:52 [INFO] - xdpw: existing screencast instance: 36 with cursor
2020/11/08 20:05:52 [DEBUG] - xdpw: matching cast instance found, but is already scheduled for destruction, skipping
2020/11/08 20:05:52 [INFO] - xdpw: screencast instance 0x558fbc979000 now has 0 references
2020/11/08 20:05:52 [INFO] - xdpw: screencast instance 0x558fbc97f020 has 1 references
2020/11/08 20:05:52 [INFO] - xdpw: 2 active screencast instances
2020/11/08 20:05:52 [INFO] - wlroots: output: eDP-1
2020/11/08 20:05:52 [INFO] - dbus: start method invoked
2020/11/08 20:05:52 [INFO] - dbus: request_handle: /org/freedesktop/portal/desktop/request/1_133/webrtc1276599495
2020/11/08 20:05:52 [INFO] - dbus: session_handle: /org/freedesktop/portal/desktop/session/1_133/webrtc_session1750199713
2020/11/08 20:05:52 [INFO] - dbus: app_id: 
2020/11/08 20:05:52 [INFO] - dbus: parent_window: 
2020/11/08 20:05:52 [DEBUG] - dbus: start: found matching session /org/freedesktop/portal/desktop/session/1_133/webrtc_session1750199713
2020/11/08 20:05:52 [DEBUG] - xdpw: destroying cast instance
2020/11/08 20:05:52 [DEBUG] - pipewire: destroying stream
2020/11/08 20:05:52 [INFO] - pipewire: stream state changed to "unconnected"
2020/11/08 20:05:52 [INFO] - pipewire: node id is -1
2020/11/08 20:05:52 [DEBUG] - pipewire: registered event 0x558fbc97df40
2020/11/08 20:05:52 [INFO] - pipewire: stream state changed to "connecting"
2020/11/08 20:05:52 [INFO] - pipewire: node id is -1
2020/11/08 20:05:52 [INFO] - pipewire: stream state changed to "paused"
2020/11/08 20:05:52 [INFO] - pipewire: node id is 30
2020/11/08 20:05:56 [INFO] - dbus: session closed
2020/11/08 20:05:56 [DEBUG] - dbus: destroying session 0x558fbc978000
2020/11/08 20:05:56 [DEBUG] - xdpw: screencast instance 0x558fbc97f020 now has 0 references
2020/11/08 20:05:56 [DEBUG] - xdpw: destroying cast instance
2020/11/08 20:05:56 [DEBUG] - pipewire: destroying stream
2020/11/08 20:05:56 [INFO] - pipewire: stream state changed to "unconnected"
2020/11/08 20:05:56 [INFO] - pipewire: node id is -1
```

**Comment by danshick** [comment]
Can you be specific about the exact browser package and version installed for those logs? Is it the `chromium` package installed by pacman or is it an AUR package like `chromium-ozone`? What does `pacman -Qi _name-of-chromium-package_` show?

How did you install xdp/xdpw? Are you building master from the AUR git package or using a packaged version?

It does appear that wayland screensharing is being attempted from your logs, but the session is being closed immediately after being started. This is very likely a pipewire negotiation issue, and logs from pipewire/chromium would be very helpful. Pipewire can be started with the environment variable `PIPEWIRE_DEBUG=3` to increase logging if necessary. Pipewire logs are noisy, and much of it is probably irrelevant. Chrome WebRTC logs probably require flags like:
```
--enable-logging --v=1 --vmodule=*/webrtc/*=2,*=-2 --enable-logging=stderr
```

If this only happens in `chromium` and it is working correctly (ignoring the mouse lag for the moment) in `fedora-firefox-wayland-bin`, I would strongly suspect this is a quirk in the specific version of chromium you have installed.

As far as mouse lag, my best guess is that frequent screencopy calls are causing a lot of CPU utilization, but I can't say for certain. Which GPU are you using? Do you see any processes consuming excessive cycles when you experience this mouse lag?

**Comment by zsolt-donca** [comment]
For this test, I am using two versions of chrome, one installed from the [extra repository](https://www.archlinux.org/packages/extra/x86_64/chromium/), and the other one is [google-chrome-dev from the AUR](https://aur.archlinux.org/packages/google-chrome-dev/). I used the former for the xwayland tests, and the latter with the flags `--enable-features=UseOzonePlatform --ozone-platform=wayland`, in wayland-native mode. I wanted to try out also `chromium-ozone`, but I don't have the time to build it.

```
❯ pacman -Qi chromium        
Name            : chromium
Version         : 86.0.4240.183-1
Description     : A web browser built for speed, simplicity, and security
Architecture    : x86_64
URL             : https://www.chromium.org/Home
Licenses        : BSD
Groups          : None
Provides        : None
Depends On      : gtk3  nss  alsa-lib  xdg-utils  libxss  libcups  libgcrypt  ttf-liberation  systemd  dbus  libpulse  pciutils  json-glib
                  desktop-file-utils  hicolor-icon-theme  icu  libxml2  fontconfig  harfbuzz  libjpeg  libpng  re2  snappy  ffmpeg  flac  libwebp
                  minizip  libxslt  freetype2  opus
Optional Deps   : pepper-flash: support for Flash content
                  libpipewire02: WebRTC desktop sharing under Wayland [installed]
                  libva: hardware-accelerated video decode [experimental] [installed]
                  kdialog: needed for file dialogs in KDE
                  org.freedesktop.secrets: password storage backend on GNOME / Xfce [installed]
                  kwallet: for storing passwords in KWallet on KDE desktops
Required By     : None
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 207.42 MiB
Packager        : Evangelos Foutras <foutrelis@archlinux.org>
Build Date      : Tue 03 Nov 2020 01:49:04 EET
Install Date    : Tue 03 Nov 2020 20:20:13 EET
Install Reason  : Explicitly installed
Install Script  : Yes
Validated By    : Signature


❯ pacman -Qi google-chrome-dev
Name            : google-chrome-dev
Version         : 88.0.4315.5-1
Description     : The popular and trusted web browser by Google (Dev Channel)
Architecture    : x86_64
URL             : https://www.google.com/chrome
Licenses        : custom:chrome
Groups          : None
Provides        : google-chrome
Depends On      : alsa-lib  gtk3  libcups  libxss  libxtst  nss
Optional Deps   : libpipewire02: WebRTC desktop sharing under Wayland [installed]
                  kdialog: for file dialogs in KDE
                  gnome-keyring: for storing passwords in GNOME keyring [installed]
                  kwallet: for storing passwords in KWallet
                  gtk3-print-backends: for printing [installed]
                  libunity: for download progress on KDE
                  ttf-liberation: fix fonts for some PDFs - CRBug #369991 [installed]
                  xdg-utils [installed]
Required By     : None
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 231.55 MiB
Packager        : Unknown Packager
Build Date      : Sun 08 Nov 2020 16:12:20 EET
Install Date    : Sun 08 Nov 2020 16:12:29 EET
Install Reason  : Explicitly installed
Install Script  : Yes
Validated By    : None
```

I also installed `xdg-desktop-portal-wlr` from the [community package](https://www.archlinux.org/packages/community/x86_64/xdg-desktop-portal-wlr/).

> Pipewire can be started with the environment variable PIPEWIRE_DEBUG=3 to increase logging if necessary.

I don't know how to start `pipewire` with this env variable set, and with access to outs standard output.  It's started, after all, by a systemd socket. Can you, please, give me some hints on this?

I captured the logs of chromium with the Chrome WebRTC logs flags:
[chromium_logs_3.txt](https://github.com/emersion/xdg-desktop-portal-wlr/files/5512345/chromium_logs_3.txt). For the logs of google-chrome-dev in ozone mode, see: [chrome_88_logs_1.txt](https://github.com/emersion/xdg-desktop-portal-wlr/files/5512352/chrome_88_logs_1.txt)

If I find out how to capture the logs of pipewire itself, then I'd gladly repeat the whole process and capture the matching logs of both pipewire and chromium/chrome, in pairs.

Regarding the laggy mouse issue, I made a screenshot of `htop` on one screen, while Firefox was doing screen sharing on my other screen: 
![screenshot_2020-11-09_19-30-31_735861790](https://user-images.githubusercontent.com/6824167/98577577-d5277780-22c4-11eb-81c8-ee4f00207dd8.png) It shows `sway`, `firefox`, `xdg-desktop-portal-wlr` being the biggest CPU users.

I also just noticed that the mouse is laggy only on the screen that is being cast (or at least I think that that is the screen that I cast, because I can't see it). The mouse seems to be moving perfectly fine on the other screen. FYI, I am using two 4K monitors (one external, and another built-in into the laptop).

**Comment by danshick** [comment]
You can kill any existing pipewire processes and start your own. As long as one is running, it won't be restarted. I'd run it like:

```
PIPEWIRE_DEBUG=3 pipewire 2>&1 | tee pipewire.log
```

Let me look through your chromium logs and I'll let you know if I need pipewire logs.

As for your mouse lag, I don't see anything terribly wrong. Still, copying 4k frames from GPU->CPU could be causing frame rate limitations. I'd be curious if this goes away if you were to change your output resolution to 1920x1080 (as a test). @emersion may know more about the performance implications. Screencopy is not the fastest thing ever, but it is all we have right now.

**Comment by danshick** [comment]
In your chromium logs, I'm seeing this:
```
[21115:21115:1109/193457.674443:INFO:base_capturer_pipewire.cc(70)] PipeWire remote state: connecting.
[21115:21115:1109/193457.674583:INFO:base_capturer_pipewire.cc(314)] PipeWire remote opened.
[E][module-protocol-native.c:562 on_remote_data()] protocol-native 0x33df5483c780: got connection error
[21115:21522:1109/193457.676427:INFO:base_capturer_pipewire.cc(73)] PipeWire remote state: unconnected.
```

I checked the pipewire 2.7 source [here](https://github.com/PipeWire/pipewire/blob/0.2.7/src/modules/module-protocol-native.c#L562) and I don't think we will get any additional info from the chromium logs, so I do think the pipewire logs will be more helpful. Can you also share `pacman -Qi` for `pipewire` and for `libpipewire02`? While I'm curious what is currently installed, I would make sure both are up to date before we continue trying to debug.

BTW, looking at both logs, they are both built with pipewire support and are both attempting to connect to a screenshare using pipewire, and are exhibiting the same issue, so I do believe either chrom{e|ium} version should work.

**Comment by danshick** [comment]
I might be able to reproduce this. Are you seeing the following in your pipewire logs?

```
[E][001643065.361131][module-protocol-native.c:305 connection_data()] protocol-native 0x55c6ab56a5b0: client 0x55c6ab65aee0 error -22 (Invalid argument)
```

**Comment by lolzballs** [comment]
I'm experiencing the same thing using the normal `google-chrome` aur package with `XDG_CURRENT_DESKTOP=sway` set. I was able to get screen sharing working in the past but after recent updates (not sure what) it no longer works. My logs appear to be similar to what @zsolt-donca posted and I do see the line that @danshick last posted.

**Comment by danshick** [comment]
> I'm experiencing the same thing using the normal `google-chrome` aur package with `XDG_CURRENT_DESKTOP=sway` set. I was able to get screen sharing working in the past but after recent updates (not sure what) it no longer works. My logs appear to be similar to what @zsolt-donca posted and I do see the line that @danshick last posted.

Thanks for confirming. I've got some testing to do.

**Comment by zsolt-donca** [comment]
I've got back with the logs. Logs from pipewire: 
[pipewire_1.log](https://github.com/emersion/xdg-desktop-portal-wlr/files/5519637/pipewire_1.log) and the matching logs from chrome 88: 
[chrome_88_logs_2.txt](https://github.com/emersion/xdg-desktop-portal-wlr/files/5519640/chrome_88_logs_2.txt). I see lots of errors with "Invalid argument" in pipewire's logs.

Regarding the versions, I am using stuff only from the official repos, and I believe I am on the latest.
```
❯ pacman -Qi pipewire
Name            : pipewire
Version         : 0.3.15-1
Description     : Server and user space API to deal with multimedia pipelines
Architecture    : x86_64
URL             : https://pipewire.org
Licenses        : LGPL2.1
Groups          : None
Provides        : libpipewire-0.3.so=0-64
Depends On      : gstreamer  gst-plugins-base  sbc  rtkit  vulkan-icd-loader  bluez-libs
                  alsa-card-profiles  libdbus-1.so=3-64  libsndfile.so=1-64  libudev.so=1-64
                  libasound.so=2-64  libsystemd.so=0-64  libglib-2.0.so=0-64  libgobject-2.0.so=0-64
Optional Deps   : pipewire-docs: Documentation
                  pipewire-jack: JACK support
                  pipewire-pulse: PulseAudio support
Required By     : fedora-firefox-wayland-bin  xdg-desktop-portal  xdg-desktop-portal-wlr
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 5.32 MiB
Packager        : Jan Alexander Steffens (heftig) <heftig@archlinux.org>
Build Date      : Wed 04 Nov 2020 23:14:39 EET
Install Date    : Fri 06 Nov 2020 22:13:55 EET
Install Reason  : Installed as a dependency for another package
Install Script  : Yes
Validated By    : Signature


❯ pacman -Qi libpipewire02
Name            : libpipewire02
Version         : 0.2.7-1
Description     : User space API to deal with multimedia pipelines (0.2)
Architecture    : x86_64
URL             : https://pipewire.org
Licenses        : LGPL2.1
Groups          : None
Provides        : libpipewire-0.2.so=1-64
Depends On      : alsa-lib  dbus
Optional Deps   : None
Required By     : None
Optional For    : chromium  google-chrome-dev  microsoft-edge-dev-bin  qt5-webengine
Conflicts With  : None
Replaces        : None
Installed Size  : 1568.36 KiB
Packager        : Jan Alexander Steffens (heftig) <jan.steffens@gmail.com>
Build Date      : Sun 08 Mar 2020 04:25:17 EET
Install Date    : Sun 26 Apr 2020 22:58:27 EEST
Install Reason  : Installed as a dependency for another package
Install Script  : No
Validated By    : Signature
```

**Comment by danshick** [comment]
@zsolt-donca Thanks, from the logs, I'm sure I can reproduce the issue now. I think a recent pipewire update may have caused issues for older consumers. I'll see what I can do.

**Comment by codicodi** [comment]
I also had this problem but it seems to be fixed with pipewire 0.3.15-2

https://github.com/archlinux/svntogit-packages/commit/f51c5ddec94332e733c333f6625cd938b11b936b

**Comment by zsolt-donca** [comment]
I can confirm that the recent (13 hours-ago) released pipewire 0.3.15-2 indeed fixes the issue for me - screen sharing now works in Chrome. I guess we can close this issue then.

I am thinking of opening another question for the mouse lag issue.

**Comment by danshick** [comment]
Fixed before I could finish finding the issue. For reference, issue is here...
https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/270
